tibanna_ffcommon/generic_qc_utils.py

```diff
@@ -1,7 +1,160 @@
+    )
+
+    if not qc_json_value:
+        raise GenericQcException(f"QC metric '{metric}' in ruleset not found in QC json")
+
+    def evaluate_target(value, operator, target):
+        if operator == ">":
+            return value > target
+        elif operator == ">=":
+            return value >= target
+        elif operator == "<":
+            return value < target
+        elif operator == "<=":
+            return value <= target
+        elif operator == "==":
+            return value == target
+        elif operator == "!=":
+            return value != target
+        else:
+            raise GenericQcException(
+                f"The ruleset contains an unsupported operator: {operator}"
+            )
+        
+    result = FAIL
+    if evaluate_target(qc_json_value.value, qc_threshold.operator, qc_threshold.pass_target):
+        result = PASS
+    elif evaluate_target(qc_json_value.value, qc_threshold.operator, qc_threshold.warn_target):
+        result = WARN
+
+    if qc_threshold.use_as_qc_flag:
+            qc_json_value.flag = result # Update individual QC flags. This will be patched to the portal
+    return result
+
+
+def evaluate_qc_ruleset(qc_json_, qc_ruleset: QC_ruleset):
+    """For given input file, calculated qc metrics and user defined ruleset,
+     returns an updated qc_json with indivifual QC flags set and the overall QC flag
+
+    Args:
+        input_wf_arg_name (string): input workflow argument that the QCs are validated for
+        qc_json_ (Dict): qc json that has been persisted as file
+        qc_ruleset_ (QC_rulesets): ruleset that is defined in the workflow
+
+    Raises:
+        GenericQcException: The overall_quality_status_rule contains metric IDs that are not defined in the rulset.
+    """
+
+    try:
+        qc_json = QC_json(**qc_json_)
+    except ValidationError as e:
+        raise GenericQcException(
+            f"The contents of the QC json file has the wrong format: {e}"
+        )
+
+    evaluated_qc_thresholds = {}
+    for qct in qc_ruleset.qc_thresholds:
+        evaluated_qc_thresholds[qct.id] = evaluate_qc_threshold(qct, qc_json)
+
+    # Insert the evaluated QC thresholds in the logical expression for the overall quality. Then evaluate it.
+    are_all_qc_pass = True # Keep track if any of the evaluated QCs is not PASS.
+    for metric_id in evaluated_qc_thresholds:
+        are_all_qc_pass = (are_all_qc_pass and (evaluated_qc_thresholds[metric_id] == PASS))
+        eqt_bool_str = "True" if evaluated_qc_thresholds[metric_id] in [PASS, WARN] else "False"
+        qc_ruleset.overall_quality_status_rule = qc_ruleset.overall_quality_status_rule.replace(f"{{{metric_id}}}", eqt_bool_str)
+
+    # Not all "{metric_id}"" have been replaced in the logical expression string. This would also fail
+    # during evaluation but we want to return a more useful error message here.
+    if "{" in qc_ruleset.overall_quality_status_rule:
```

Comment:
Is this really guaranteed? Why not check for `{metric_id}` explicitly to ensure you're not picking up something else?

Reply:
I think this is fine. If there is anything that hasn't been replace there must be a `{` in the status rule. If the status rule itself has the wrong syntax, e.g., `{c1} and c2}`, it will fail later during evaluation. This error is supposed to catch syntactically valid status rules where one of the metric IDs has not been defined.